### PR TITLE
Ab/hashing

### PR DESCRIPTION
**Pull request by aborochoff**
Hashing resource identifier is expensive due to allocating a 4 index array for hash calculation.  Use the concat'd string hash code instead which caches.  Also only stores the single toString string and returns the sub-components from the main resource identifier string.

### CODE REVIEW

File: src/main/java/com/palantir/ri/ResourceIdentifier.java [hunk 1]
```diff
@@ -59,16 +59,25 @@
...
+    private final int instanceIndex;
+    private final int typeIndex;
+    private final int locatorIndex;
 
     private ResourceIdentifier(String service, String instance, String type, String locator) {
-        this.service = service;
-        this.instance = instance == null ? "" : instance;
-        this.type = type;
-        this.locator = locator;
+        String safeInstnace = instance == null ? "" : instance;
```
Comment (bavardage) [inline src/main/java/com/palantir/ri/ResourceIdentifier.java]: typo

**Comment by markelliot** [comment]
why not memoize and additionally store the string rather than the indexes?

**Comment by aborochoff** [comment]
I didn't want to double the memory footprint of RID.  I think get locator is the only one commonly called component, I don't think its on any perf critical code paths, so I made the trade off.

File: src/main/java/com/palantir/ri/ResourceIdentifier.java [hunk 2]
```diff
@@ -59,16 +59,25 @@
...
+                + service + SEPARATOR
+                + safeInstance + SEPARATOR
+                + type + SEPARATOR
+                + locator;
+
+        serviceIndex = RID_CLASS.length() + SEPARATOR.length() + service.length();
+        instanceIndex = serviceIndex + SEPARATOR.length() + safeInstance.length();
+        typeIndex = instanceIndex + SEPARATOR.length() + type.length();
+        locatorIndex = typeIndex + SEPARATOR.length() + locator.length();
+
```
Comment (uschi2000) [inline src/main/java/com/palantir/ri/ResourceIdentifier.java]: nit: extra \n

**Comment by uschi2000** [comment]
I'm fine with this. Have you checked that we have code coverage for this?

**Comment by aborochoff** [comment]
https://github.com/palantir/resource-identifier/blob/develop/src/test/java/com/palantir/ri/ResourceIdentifierTest.java#L146 covers the changes that I made and makes sure that the sub components come out the way they went in.

Just made a slight change to the equals method to use the native equals on the string instead of Object.equals as well as fixing the nit.

**Review by uschi2000 — APPROVED** [review]

File: src/main/java/com/palantir/ri/ResourceIdentifier.java [hunk 3]
```diff
@@ -104,43 +113,39 @@ public String getType() {
      * @return the locator component from this identifier
      */
     public String getLocator() {
-        return locator;
+        CharBuffer.wrap(resourceIdentifier);
```
Comment (j-baker) [inline src/main/java/com/palantir/ri/ResourceIdentifier.java:116]: ?

Comment (aborochoff) [reply src/main/java/com/palantir/ri/ResourceIdentifier.java:116]: Left over from some perf testing 😞

I’ll fix when I’m back in the office.

**Pull request merged**